src/System.Private.CoreLib/shared/System/Collections/ObjectModel/Collection.cs

```diff
@@ -162,6 +170,61 @@ protected virtual void SetItem(int index, T item)
             items[index] = item;
         }
 
+        protected virtual void InsertItemsRange(int index, IEnumerable<T> collection)
+        {
+            if (collection == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.list);
+            }
+
+            if (items.IsReadOnly)
+            {
+                ThrowHelper.ThrowNotSupportedException(ExceptionResource.NotSupported_ReadOnlyCollection);
+            }
+
+            if ((uint)index > (uint)items.Count)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.index, ExceptionResource.ArgumentOutOfRange_ListInsert);
+            }
+
+            int i = 0;
+            foreach (T item in collection)
+            {
+                Insert(index + i, item);
```

Comment:
And if we wanted, this could be optimized to avoid the virtual calls to `InsertItem` for each item, when the instance is not a subclass and when `items` is `List<T>`, something like:

```c#
if (GetType() == typeof(Collection<T>) && items is List<T> list)
{
    list.InsertRange(index, collection);
}
else
{
    foreach (T item in collection)
    {
        InsertItem(index++, item);
    }
}
```

Reply:
@justinvp this is great feedback and suggestions, thanks for adding these notes here. We want this code to be as lean as possible. I just updated it locally with your suggestions and all tests are passing, I'll be updating the PR shortly